homeassistant/components/ads/__init__.py

```diff
@@ -222,37 +255,56 @@ def add_device_notification(self, name, plc_datatype, callback):
     def _device_notification_callback(self, notification, name):
         """Handle device notifications."""
         contents = notification.contents
-
         hnotify = int(contents.hNotification)
         _LOGGER.debug("Received notification %d", hnotify)
 
-        # get dynamically sized data array
+        # Get dynamically sized data array
         data_size = contents.cbSampleSize
-        data = (ctypes.c_ubyte * data_size).from_address(
+        data_address = (
             ctypes.addressof(contents)
             + pyads.structs.SAdsNotificationHeader.data.offset
         )
+        data = (ctypes.c_ubyte * data_size).from_address(data_address)
 
-        try:
-            with self._lock:
-                notification_item = self._notification_items[hnotify]
-        except KeyError:
+        # Acquire notification item
+        with self._lock:
+            notification_item = self._notification_items.get(hnotify)
+
+        if not notification_item:
             _LOGGER.error("Unknown device notification handle: %d", hnotify)
             return
 
-        # Parse data to desired datatype
-        if notification_item.plc_datatype == pyads.PLCTYPE_BOOL:
-            value = bool(struct.unpack("<?", bytearray(data))[0])
-        elif notification_item.plc_datatype == pyads.PLCTYPE_INT:
-            value = struct.unpack("<h", bytearray(data))[0]
-        elif notification_item.plc_datatype == pyads.PLCTYPE_BYTE:
-            value = struct.unpack("<B", bytearray(data))[0]
-        elif notification_item.plc_datatype == pyads.PLCTYPE_UINT:
-            value = struct.unpack("<H", bytearray(data))[0]
-        elif notification_item.plc_datatype == pyads.PLCTYPE_DINT:
-            value = struct.unpack("<i", bytearray(data))[0]
-        elif notification_item.plc_datatype == pyads.PLCTYPE_UDINT:
-            value = struct.unpack("<I", bytearray(data))[0]
+        # Data parsing based on PLC data type
+        plc_datatype = notification_item.plc_datatype
+        unpack_formats = {
+            pyads.PLCTYPE_BOOL: "<?",
+            pyads.PLCTYPE_BYTE: "<b",
+            pyads.PLCTYPE_INT: "<h",
+            pyads.PLCTYPE_UINT: "<H",
+            pyads.PLCTYPE_SINT: "<b",
+            pyads.PLCTYPE_USINT: "<B",
+            pyads.PLCTYPE_DINT: "<i",
+            pyads.PLCTYPE_UDINT: "<I",
+            pyads.PLCTYPE_WORD: "<H",
+            pyads.PLCTYPE_DWORD: "<I",
+            pyads.PLCTYPE_REAL: "<f",
+            pyads.PLCTYPE_LREAL: "<d",
+            pyads.PLCTYPE_STRING: "string",
+            pyads.PLCTYPE_TIME: "<i",
+            pyads.PLCTYPE_DATE: "<i",
+            pyads.PLCTYPE_DT: "<i",
+            pyads.PLCTYPE_TOD: "<i",
+        }
+
+        if plc_datatype in unpack_formats:
```

Comment:
I think we can make `string` a separate case, maybe the bool should be as well. the goal with my last comment was to reduce repetition and that has been achieved, but I think we overshot a bit

Reply:
yes, now i handle `bool` and `string` a special way.